### PR TITLE
feat: add ln and log10 numeric functions

### LINKED ISSUE #211: Feature Request: Add ln (natural logarithm) and log10 Scalar Functions
### Overview of feature request

We need to expand the numeric scalar functions available in Drasi. This task involves implementing the natural logarithm (`ln`) and base-10 logarithm (`log10`) functions in the core engine and registering them for both Cypher and GQL query languages.

This is a great task for new contributors to understand how Drasi's function evaluation system works.

### Acceptance criteria

1. **Implement Core Logic**:
   - Create `Ln` and `Log10` structs in `core/src/evaluation/functions/numeric/numeric.rs`.
   - Implement the `ScalarFunction` trait for both structs.
   - **Behavior**:
     - Accept a single numeric argument (`Integer` or `Float`).
     - Return `Null` if the input is `Null`.
     - Return a `Float` result.
     - Return an appropriate error if the input is non-numeric.
     - Handle domain errors (inputs <= 0) gracefully (either return error or null, consistent with other math functions).

2. **Register Functions**:
   - Register the new functions in `functions-cypher/src/lib.rs` (as `ln` and `log10`).
   - Register the new functions in `functions-gql/src/lib.rs` (as `ln` and `log10`).

3. **Testing**:
   - Add unit tests to verify behavior.
   - Cover edge cases: positive integers, positive floats, zero, negative numbers, and null.

### Additional context

- Look at `core/src/evaluation/functions/numeric/numeric.rs` for examples of existing numeric functions like `Abs`, `Ceil`, `Floor`.
- Look at `functions-cypher/src/lib.rs` and `functions-gql/src/lib.rs` to see how functions are registered.
- The `ScalarFunction` trait is defined in `core/src/evaluation/functions/mod.rs`.
- Remember to handle `VariableValue::Integer` and `VariableValue::Float` variants.

### PR DESCRIPTION
**Pull request by IronJam11**
## Description

This pull request implements two mathematical logarithmic functions for the Drasi query evaluation system: `ln()` (natural logarithm) and `log10()` (base-10 logarithm).

### New Functions
- **`ln(x)`** - Returns the natural logarithm (base e) of a numeric value
- **`log10(x)`** - Returns the base-10 logarithm of a numeric value

Both functions:
- Accept integer or float inputs
- Return float values
- Handle null inputs (return null)
- Validate domain (error for values ≤ 0)
- Validate argument count and types

### Files Added

**Implementation:**
- `core/src/evaluation/functions/numeric/ln.rs`
- `core/src/evaluation/functions/numeric/log10.rs`

**Tests:**
- `core/src/evaluation/functions/numeric/tests/ln_tests.rs`
- `core/src/evaluation/functions/numeric/tests/log10_tests.rs`

**Modified:**
- `core/src/evaluation/functions/numeric/tests.rs` - Added test module declarations

### Test Coverage
- Argument validation (count and type checking)
- Positive integer and float inputs
- Domain error handling (negative values and zero)
- Null input handling
- Type mismatch validation

## Type of change

- This pull request adds or changes features of Drasi and has an approved issue (issue link required).

Fixes #211

### CODE REVIEW

File: core/src/evaluation/functions/numeric/ln.rs [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+use crate::evaluation::functions::ScalarFunction;
+use crate::evaluation::variable_value::float::Float;
+use crate::evaluation::variable_value::VariableValue;
+use crate::evaluation::{ExpressionEvaluationContext, FunctionError, FunctionEvaluationError};
+
+#[derive(Debug)]
+pub struct Ln {}
+
+#[async_trait]
+impl ScalarFunction for Ln {
```
Comment (ruokun-niu) [inline core/src/evaluation/functions/numeric/ln.rs]: In both **Neo4j Cypher** and the **openCypher** specification, the natural logarithm (base *e*) is named **`log()`**, not `ln()`:

- Neo4j docs: https://neo4j.com/docs/cypher-manual/current/functions/mathematical-logarithmic/#functions-log  
  → `log(input)` = natural logarithm  
  → `log10(input)` = base-10 logarithm

- openCypher spec (section on mathematical functions) follows the same convention.

File: core/src/evaluation/functions/numeric/ln.rs [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+                let value = match n.as_i64() {
+                    Some(n) => n as f64,
+                    None => {
+                        return Err(FunctionError {
+                            function_name: expression.name.to_string(),
+                            error: FunctionEvaluationError::OverflowError,
+                        })
+                    }
+                };
+                if value <= 0.0 {
```
Comment (ruokun-niu) [inline core/src/evaluation/functions/numeric/ln.rs]: This is actually not valid. log(0) should return -infinity. 

That being said, I don't think drasi-core currently has support for infinity in our internal data type. I can point to a few areas if you are interested in adding this support

File: core/src/evaluation/functions/numeric/ln.rs (hunk 1 above)
Comment (IronJam11) [reply core/src/evaluation/functions/numeric/ln.rs]: Should i change it to Log instead of Ln then? @ruokun-niu

File: core/src/evaluation/functions/numeric/ln.rs (hunk 2 above)
Comment (IronJam11) [reply core/src/evaluation/functions/numeric/ln.rs]: Sure

File: core/src/evaluation/functions/numeric/ln.rs (hunk 1 above)
Comment (ruokun-niu) [reply core/src/evaluation/functions/numeric/ln.rs]: Yes please, thanks!

File: core/src/evaluation/functions/numeric/ln.rs (hunk 2 above)
Comment (ruokun-niu) [reply core/src/evaluation/functions/numeric/ln.rs]: So the internal data types for drasi core are defined in this folder called [variable_value](https://github.com/drasi-project/drasi-core/tree/main/core/src/evaluation/variable_value). You can review the implementation of the numeric values in `integer.rs` and `float.rs`. The code in `mod.rs` is also very important and useful. 

Please take a look at these files and let me know what you think! If you think this is beyond the scope of this PR please let me know and I am happy to create a separate issue for someone else to take on this.

Comment (IronJam11) [reply core/src/evaluation/functions/numeric/ln.rs]: I incorporated the negative infinity as you requested

**Comment by IronJam11** [comment]
Screenshot 2026-02-07 at 12 01 09 AM

File: core/src/evaluation/functions/numeric/ln.rs (hunk 1 above)
Comment (IronJam11) [reply core/src/evaluation/functions/numeric/ln.rs]: Done

File: core/src/evaluation/functions/numeric/ln.rs (hunk 2 above)
Comment (IronJam11) [reply core/src/evaluation/functions/numeric/ln.rs]: The float implementation checks for is_nan instead of is_finite now to accomodate f64::NEG_INFINITY

**Comment by ruokun-niu** [comment]
Thanks @IronJam11 ! This looks solid.

**Review by ruokun-niu — APPROVED** [review]

**Review by danielgerlag — COMMENTED** [review]
are we registering these functions?

**Comment by IronJam11** [comment]
Yes, it gets registered and can be used as shown in the tests

File: core/src/evaluation/variable_value/float.rs [hunk 3]
```diff
@@ -69,7 +69,7 @@ impl Float {
 
     pub(crate) fn as_f64(&self) -> Option<f64> {
         match self.value {
-            n if n.is_finite() => Some(n),
+            n if !n.is_nan() => Some(n),
```
Comment (danielgerlag) [inline core/src/evaluation/variable_value/float.rs:72]: How are we going to represent infinity in the query output?
We need to add some e2e integration tests (with actual queries) around this, that also exercises serialization/persistence, deserialization and shows how infinite numbers are represented on query results / downstream reactions, etc.

**Review by danielgerlag — CHANGES_REQUESTED** [review]

File: core/src/evaluation/variable_value/float.rs (hunk 3 above)
Comment (IronJam11) [reply core/src/evaluation/variable_value/float.rs:72]: Valid, we need to ensure that the float::infinity runs smoothly throughout the system without panicking.

Comment (IronJam11) [reply core/src/evaluation/variable_value/float.rs:72]: I have been looking at the issue for a day, and I have a solution. The async test test_infinity_e2e2_source_query_reaction() exercises the full pipeline (refer to the latest commit) that is: 

ApplicationSource -> Query(cypher) -> Reaction (SSE) -> JSON serialisation 

The issue appeared in the full Source → Query → Reaction → JSON flow. When the query computed log10(0) or log(0), the result was -Infinity. At the core level, we already had a defined rule that non-finite floats must serialize as null. That rule was implemented in multiple conversion layers, but the SSE reaction path was not using those conversions.

In the flow, the Query produces a VariableValue containing a Float. During reaction processing, manager.rs calls convert_variable_value_to_json. Instead of using the existing From<VariableValue> conversion (which checks finiteness), it directly calls to_string() on the Float, producing "-inf". This changed the type from a numeric value to a string and violated our JSON contract.

The changes ensure consistency across the pipeline:

1. In mod.rs, From<VariableValue> for serde_json::Value enforces the non-finite → null policy.
2. In element_value.rs, From<&ElementValue> routes through the same safe conversion.
3. In float.rs, the guarded From<Float> for Number prevents accidental serialization of non-finite values.
4. In manager.rs, convert_variable_value_to_json now checks finiteness instead of calling to_string().

With this, the entire flow, from query evaluation to SSE emission, consistently serializes non-finite floats as null, preserving type correctness and JSON validity.

Testing using:  cargo test -p drasi-reaction-sse test_infinity_e2e -- --nocapture

File: core/src/evaluation/functions/numeric/log.rs [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+use drasi_query_ast::ast;
+
+use crate::evaluation::functions::ScalarFunction;
+use crate::evaluation::variable_value::float::Float;
+use crate::evaluation::variable_value::VariableValue;
+use crate::evaluation::{ExpressionEvaluationContext, FunctionError, FunctionEvaluationError};
+
+#[derive(Debug)]
+pub struct Log {}
+
```
[the remark this answers is not in the document]
Comment (ruokun-niu) [reply core/src/evaluation/functions/numeric/log.rs:25]: Resolving as this is the correct name according to openCypher

**Comment by ruokun-niu** [comment]
@IronJam11 Can you address the above comments?

**Comment by ruokun-niu** [comment]
Hi @IronJam11 are you still working on this?

**Comment by ruokun-niu** [comment]
Hi @IronJam11 I am going to close this PR
